src/python/pants/core/goals/lint.py

```diff
@@ -36,41 +38,90 @@ class InvalidLinterReportsError(Exception):
 
 @dataclass(frozen=True)
-class LintResult:
+class LintResult(EngineAware):
     exit_code: int
     stdout: str
     stderr: str
-    linter_name: str
     report: Optional[LintReport]
 
-    @staticmethod
+    @classmethod
     def from_fallible_process_result(
+        cls,
         process_result: FallibleProcessResult,
         *,
-        linter_name: str,
         strip_chroot_path: bool = False,
         report: Optional[LintReport] = None,
     ) -> "LintResult":
         def prep_output(s: bytes) -> str:
             return strip_v2_chroot_path(s) if strip_chroot_path else s.decode()
 
-        return LintResult(
+        return cls(
             exit_code=process_result.exit_code,
             stdout=prep_output(process_result.stdout),
             stderr=prep_output(process_result.stderr),
-            linter_name=linter_name,
             report=report,
         )
 
 
-class LintResults(Collection[LintResult]):
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class LintResults(EngineAware):
     """Zero or more LintResult objects for a single linter.
 
     Typically, linters will return one result. If they no-oped, they will return zero results.
     However, some linters may need to partition their input and thus may need to return multiple
     results. For example, many Python linters will need to group by interpreter compatibility.
     """
 
+    results: Tuple[LintResult, ...]
+    linter_name: str
+
+    def __init__(self, results: Iterable[LintResult], *, linter_name: str) -> None:
+        self.results = tuple(results)
+        self.linter_name = linter_name
+
+    @property
+    def skipped(self) -> bool:
+        return bool(self.results) is False
+
+    @memoized_property
+    def exit_code(self) -> int:
+        return next((result.exit_code for result in self.results if result.exit_code != 0), 0)
+
+    @memoized_property
+    def reports(self) -> Tuple[LintReport, ...]:
+        return tuple(result.report for result in self.results if result.report)
+
+    def level(self) -> Optional[LogLevel]:
+        if self.skipped:
+            return LogLevel.DEBUG
+        return LogLevel.WARN if self.exit_code != 0 else LogLevel.INFO
+
+    def message(self) -> Optional[str]:
+        if self.skipped:
+            return "skipped."
+        message = "succeeded." if self.exit_code == 0 else "failed."
+
+        def msg_for_result(result: LintResult) -> str:
+            msg = ""
+            if result.stdout:
+                msg += f"\n{result.stdout}"
+            if result.stderr:
+                msg += f"\n{result.stderr}"
+            if msg:
+                msg = f"{msg.rstrip()}\n\n"
+            return msg
+
+        if len(self.results) == 1:
+            results_msg = msg_for_result(self.results[0])
+        else:
+            results_msg = "\n"
+            for i, result in enumerate(self.results):
+                msg = msg_for_result(result) or "\n"
+                results_msg += f"Partition #{i + 1}:{msg}"
```

Comment:
Yeah @benjyw and I struggled with how to represent partitions, and they have never been represented well. 

Right now, partitioning only happens for Flake8, Bandit, and Pylint when you have different interpreter constraints across your repo. For example, if you have a Python 2 part of your codebase and a Python 3 part of your codebase, we will partition so that the interpreter constraints don't conflict.

Maybe the solution is for `LintResult` to have an optional field called `partition_description`? Then we could say something like this:

```
Partition #1 - CPython==2.7.*:

Partition #2 - CPython>=3.6:
failed for foo..
```

Thoughts?

Reply:
Done. Lmk if you have suggestions to improve this.

Reply:
The `Partition #X - ` part is still out of the blue it seems from a user perspective. Would there be something confusing, misleading or jarring about just concatenating all error output and eliding all success output here? Concretely sticking with this example, just collapse to:
```
failed for foo...
```

Reply:
I think so. Your example works well for Flake8, but imagine if we were partitioning Black. Black always outputs a message, so you would get something like:

```
<cake>
Success! 8 files would not be formatted.

<sad emoji>
Failure! 3 files would be formatted.
```

Compared to:


```
Partition #1 - ['CPython>=3.7']:
<cake>
Success! 8 files would not be formatted.

Partition #2 - ['CPython==2.7.*']:
<sad emoji>
Failure! 3 files would be formatted.
```

--

We discuss this batching in https://www.pantsbuild.org/v2.0/docs/python-lint-goal, and I suspect it's something we want to be transparent that we're doing.

--

I could be convinced otherwise, though. This is definitely still weird.

Reply:
We're not partitioning Black though, so maybe this isn't something we need to worry about. 

Also, in the future we might want to partition runs of various tools in various ways for performance reasons (we've talked about this), in which case we'll have to find some way to merge results for display anyway.

Reply:
> We're not partitioning Black though

True, we're not. But we are partitioning Pylint and Bandit. Both of those produce a summary at the end, like `Your code has been rated at 7.14/10`. It's confusing imo to have that summary with no context on why it's showing up twice, as it's not immediately obvious that Pants is partitioning things for you based on interpreter constraints.

I updated the wording for this partition message, and also updated the PR description to show how it looks with both Flake8 and Pylint.